Feature/signup

WonHeejin opened the pull request:
커스텀 예외처리 클래스 생성(/global/exception/...)
암호화 관련 클래스(Salt.java 등)는 SprigSecurity 도입 시 삭제될 예정(SpringSecurity에서 제공하는 기능으로 변경)

WonHeejin committed:
build: guava, h2 의존성 추가

guava -> 암호화 salt 생성을 위한 의존성
h2 -> 테스트코드 작성 시 인메모리 db 사용

WonHeejin committed:
feat:사용자 엔티티 생성

WonHeejin committed:
feat:암호화 기능 생성

WonHeejin committed:
test: 사용자 생성, 암호화 기능 테스트 코드 작성

WonHeejin committed:
feat: 커스텀 예외처리 클래스 생성

WonHeejin committed:
feat: 회원가입 기능 생성

WonHeejin committed:
test: 회원가입 기능 테스트

hrJin9 commented on src/test/java/com/tlog/backend/member/MemberTest.java:
테스트 작성할때 Mock 객체도 이용해보는게 좋을 것 같아요! 
그리고 리파지토리 테스트로 보이는데 혹시 서비스 테스트도 진행하실건가요?

WonHeejin replied:
MemberTest.java에서는 리파지토리 테스트를 진행하고, MemberControllerTest.java에서 Mock 객체를 이용한 통합테스트 진행했습니다. 서비스단은 컨트롤러에서 통합테스트로 대체하였는데 따로 테스트를 구현하는게 좋을까요?

hrJin9 replied:
넵넵 통합테스트도 좋지만 서비스에서 @MockBean을 이용해서 서비스단위테스트를 진행해보는 것도 좋을 것 같아요!

WonHeejin replied:
repository 테스트같은 경우는 @MockBean으로 의존성을 추가하게 되는 경우 실제로는 동작하지 않기 때문에 마지막에 값이 잘 저장되었는지 확인하는 코드가 null로 나옵니다. 다른 사람들의 깃허브 코드를 봐도 리포지토리 테스트는 @Autowired를 사용하네요.

`assertThat(savedMember.getId()).isEqualTo(savedSalt.getMember().getId()); //salt가 해당 멤버에 맞게 저장됐는지 확인`
> java.lang.NullPointerException: Cannot invoke "com.tlog.backend.member.domain.Member.getId()" because "savedMember" is null

hrJin9 replied:
넵~ 서비스 단위 테스트를 진행해주시면 될것같아요!

hrJin9 commented on src/main/java/com/tlog/backend/member/web/dto/MemberSignUpRequest.java:
```diff
@@ -0,0 +1,40 @@
+	@NotBlank(message = "�г����� �Է����ּ���.")
+	private String nickname;
+	
+	@Builder
+	public MemberSignUpRequest(String email, String password, String checkPassword, String nickname) {
+		this.email = email;
+		this.password = password;
+		this.checkPassword = checkPassword;
+		this.nickname = nickname;
+	}
```
혹시 이부분 Builder 패턴 쓰신 이유가 있을까요? Request 값을 받는 부분에 변경이 필요한지..!?

WonHeejin replied:
테스트 진행 시 테스트 데이터를 입력하기 위해 추가했습니다! 혹시 Builder를 사용하지 않고 테스트를 진행할 방법이 있을까요?

hrJin9 replied:
아항 테스트용도였구나! 보통 Request 값은 변경이 필요 없으니까 쓴 이유가 있나 하고 물어본거였어요! 
제 생각엔 개인의 스타일이라서 사용해도 문제 없을 것 같아요.
아니면 커스텀빌더니까 빌더메서드이름을 지정해주면 유지보수하는 입장에서 더 알아보기 쉬울것 같기도 해요!

hrJin9 commented on src/main/java/com/tlog/backend/member/web/dto/MemberSignUpRequest.java:
```diff
@@ -0,0 +1,40 @@
+		this.nickname = nickname;
+	}
+	
+	public Member toEntity() {
+		return Member.builder()
+				.email(email)
+				.nickname(nickname)
+				.password(password)
+				.role(Role.GUEST)
+				.build();
+	}
```
제 의견으로는 Request는 Presentation 레이어기 때문에 domain을 알 필요는 없을 것 같아요!
그리고 Role을 GUEST로 넣어주는 로직이 있는데, 이건 비즈니스(도메인) 로직이 아닐까하는 생각이 드네욥..!

WonHeejin replied:
아래와 같이 서비스단에서 리퀘스트를 엔티티로 변환할때 편리하게 사용하려고 만들었는데, 레이어별 역할 구분이 모호해지는 문제가 있는것 같습니다. 이 부분은 수정해서 재커밋 할게요!
`//회원 정보 저장
Member savedMember = repository.save(request.toEntity());`

WonHeejin committed:
refactor: request to entity 수행 위치 이동

request는 presentation 레이어이기 때문에 domain을 알 필요 없음.
게다가 role에 Gest를 저장하는 로직은 비즈니스레이어의 역할.
따라서 Service단에서 해당 로직 수행하도록 변경함.

WonHeejin committed:
test: 테스트용 application.yml이 안불러와지는 문제 해결

WonHeejin committed:
style: request의 builder 변수명 수정

request에 존재하는 builder는 테스트 시 데이터를 생성하는 목적으로만 사용되므로 유지보수 시 알아볼 수 있도록 변수명
수정, 주석 추가함.

WonHeejin committed:
style: 빌더 이름 추가

WonHeejin commented on src/main/java/com/tlog/backend/member/service/MemberService.java:
```diff
@@ -0,0 +1,57 @@
+		}
+		
+		//ȸ�� ���� ����
+		Member entity = Member.builder()
+				.email(request.getEmail())
+				.password(request.getPassword())
+				.nickname(request.getNickname())
+				.role(Role.GUEST)
+				.build();
```
request에 있던 toEntity() 삭제하고 서비스단에서 처리하도록 수정했습니다.

WonHeejin commented on src/main/java/com/tlog/backend/member/web/dto/MemberSignUpRequest.java:
```diff
@@ -0,0 +1,29 @@
+	@NotBlank(message = "�г����� �Է����ּ���.")
+	private String nickname;
+	
+	@Builder(builderMethodName = "testDataBuilder") //�׽�Ʈ ������ ������ ����
+	public MemberSignUpRequest(String test_email, String test_password, String test_checkPassword, String test_nickname) {
+		this.email = test_email;
+		this.password = test_password;
+		this.checkPassword = test_checkPassword;
+		this.nickname = test_nickname;
+	}
```
빌더명 추가했습니다.

hrJin9 reviewed (approved):
굿굿 수고하셨습니다~~
